Add Quarkus Data Hibernate getting started tutorial (#2784)
quarkusio/quarkusio.github.io: _guides/quarkus-data-hibernate-getting-started.adoc

include::{includes}/extension-status.adoc[]

== Overview

Database access is something almost every application needs, whether it's persisting data or querying tables.
In this guide we start from scratch and build up a repository for an online bookstore, going from your first SQL query to compile-time validated repositories and reactive database access. By the end you'll have a solid understanding of how data access works in Quarkus.

== Prerequisites

To follow this tutorial you need:

- Java 17+ installed (check with `java -version`)
- The Quarkus CLI installed (see the xref:cli-tooling.adoc[CLI guide] for instructions)
- A container runtime (Docker, Podman, or similar) running, so that Quarkus Dev Services can start a PostgreSQL container for you automatically

== Introduction

You're working on a physical bookstore website, in which you want users to be able to browse books with titles.
In this tutorial, we won't deal with the frontend and the user interface, we just want to connect to the database.

We'll start this tutorial by creating a new empty Quarkus application. The easiest way to do so is to have the Quarkus CLI tool available:

[source,bash]
----

> quarkus create app --name="Quarkus Data Tutorial" quarkus-data-tutorial --platform-bom=io.quarkus.platform:quarkus-bom:3.37.0.CR1

Reviewer: I don't think we should hardcode versions there, it will not age well.

> ----

Reviewer 2: There are asciidoc "partials" that you can include for this, and will display a nice UI where you can chose between the CLI, Maven or Gradle.

E.g. https://quarkus.io/guides/hibernate-orm#setting-up-and-configuring-hibernate-orm

<img width="1272" height="634" alt="Image" src="https://github.com/user-attachments/assets/d61d68e9-5bce-4c86-9ae5-70aeff3bd9c0" />

Or https://quarkus.io/guides/getting-started#bootstrapping-the-project

<img width="1272" height="634" alt="Image" src="https://github.com/user-attachments/assets/5590324c-b723-48c9-a090-bebf1bfbcd43" />

Maybe use something similar here?

Let's start by importing the extensions:

[source,bash]
----
cd quarkus-data-tutorial

quarkus extension add quarkus-data-hibernate

> quarkus extension add quarkus-jdbc-postgresql

Reviewer 2: Some comments, take what you like:

1. Same as above, you might want to present CLI/Maven/Gradle
2. You can add extensions directly in the first CLI call (`quarkus create app`), no need for a separate call
3. If you really want to split this, you can probably use a single CLI call to add both extensions, which would likely be a bit faster. Not sure if extensions are comma-separated or space-separated though.

----

The first dependency, `quarkus-data-hibernate`, is the Quarkus extension that provides everything you need for database access.

> The second dependency, `quarkus-jdbc-postgresql`, is the JDBC driver that tells Quarkus how to talk to a PostgreSQL database. If you're using a different database (MySQL, MariaDB, H2, etc.), pick the corresponding `quarkus-jdbc-*` driver instead. Quarkus Data Hibernate cannot work without a JDBC driver.

Reviewer 2: A link to the relevant section of the datasource guide that lists available extensions would likely make sense here?

That's all you need to start querying your database.

Quarkus Data Hibernate uses the Hibernate annotation processor, a key component that reads your code at compile time, validates queries, and generates implementation classes under `target/generated-sources/annotations/`.

> These generated classes are compiled together with your code and wired up automatically by Quarkus. They are also simple to read, in case you want to understand how it works internally.

Reviewer 2: Not sure this path holds for Gradle.

```suggestion
Quarkus Data Hibernate uses the Hibernate annotation processor, a key component that reads your repository interfaces at compile time, validates queries, and generates implementation classes.
These generated classes are compiled together with your code and wired up automatically by Quarkus. They are also simple to read, in case you want to understand how it works internally.
```

Reviewer: I don't know. Perhaps here it's too early, but somewhere, it would be useful to tell people where those files are generated.

> To enable it, find the `maven-compiler-plugin` in your `pom.xml` (it was generated by `quarkus create app`) and add the annotation processor configuration to it:

Reviewer 2: Also @FroMage do you think `quarkus ext add` could add the processor automatically? I think you looked into this when working on your PR to run processing even if this configuration is absent, no?

Reviewer: I don't think there's machinery for that, not unless it includes the quickstarts code.

Perhaps @aloubyansky or @ia3andy know?

Reviewer: It would be nice.

Author: https://github.com/quarkusio/quarkus/issues/55651

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<plugin>
    <artifactId>maven-compiler-plugin</artifactId>
    <version>${compiler-plugin.version}</version>
    <configuration>
        <parameters>true</parameters> <!-- already present -->
        <annotationProcessorPathsUseDepMgmt>true</annotationProcessorPathsUseDepMgmt>
        <annotationProcessorPaths>
            <path>
                <groupId>org.hibernate.orm</groupId>
                <artifactId>hibernate-processor</artifactId>
            </path>
        </annotationProcessorPaths>
    </configuration>
</plugin>

> ----

Reviewer 2: Needs Gradle example?

Now run Quarkus continuous testing in a separate tab of your terminal:

[source,bash]
----
quarkus test
----

> You don't need to restart Quarkus testing, it will re-run tests continuously after every change.

Reviewer 2: Is the use of `quarkus test` instead of `quarkus dev` an explicit choice? I've always felt that dev mode, with the swagger UI, was particularly useful to try things out, but it's true that continuous testing is an interesting approach, too.

Author: it fits the tutorial in which we just need to see if everything works and to see compiler errors when repositories are wrong

INSERT INTO Book (id, title, isbn, price, publisher_id) VALUES (6, 'Head First Design Patterns', '978-0596007126', 38.00, 3);
INSERT INTO Book (id, title, isbn, price, publisher_id) VALUES (7, 'Java Concurrency in Practice', '978-0321349606', 48.00, 4);
INSERT INTO Book (id, title, isbn, price, publisher_id) VALUES (8, 'Kotlin in Action', '978-1617293290', 44.00, 4);
----

This SQL script creates sequences for ID generation, the `Publisher` and `Book` tables, and inserts sample data.
Each book is linked to a publisher.
The sequences start past the imported data so that new IDs don't collide.
While developing, Quarkus imports this SQL file automatically at startup, so in this tutorial we have something to start with.

Quarkus Data Hibernate can handle the database schema for you.
Since our `import.sql` already handles the full schema creation, let's tell Hibernate not to manage it and tell Dev Services to run our script when the database container starts.
Add this to `src/main/resources/application.properties`:

[source,properties]
----
quarkus.hibernate-orm.schema-management.strategy=none
quarkus.datasource.devservices.init-script-path=import.sql

> ----

Reviewer 2: As discussed, I think this can be a bit overwhelming for a getting started tutorial that's supposed to introduce the basics, especially in the first steps. I think there are things to improve here -- either in your tutorial or in Quarkus -- because it's clearly not smooth.

First, the schema could be generated by Hibernate, which would be a much better experience for development. But then when you start your tutorial, you don't have entities, so that forces you do all this -- including overriding defaults with `quarkus.hibernate-orm.schema-management.strategy=none`, even though the default of `drop-and-create` is there precisely to help!
One solution to this would be to split this into two separate guides, BTW: one for pure SQL and one for mapping. It could make sense because the audience is not the same: people doing entity mapping will use SQL only in very specific edge cases, while people using SQL all over the place will not want to even hear about Hibernate. Also, people using SQL all over the place will be much more comfortable with a big DDL script at the start of the tutorial.

Second, this is mixing schema and data population, which in actual applications (when going further) would be separate concerns. It'd be nice if we clearly separated them here. Unfortunately until we solve #21866 , the only solution to separate data population would be to move it to the test directly.
Though if things stay simple, creating the data in the test is an idea too: you could just have a `@BeforeAll` method that calls the repository to add instances?

Reviewer: Yes, agreed. This is scary as hell, and only required for freaks like @maxandersen who are allergic to `@Entity`, let's split this off into a file for Max please :)

Author: Moved to the SQL tutorial

> TIP: JDBC connections are called "Data Sources" in Quarkus. In dev mode, if you don't configure a datasource, Quarkus will automatically start a PostgreSQL container for you via Dev Services. This means you can run `quarkus dev` and start coding without installing or configuring a database. See the xref:datasource.adoc[Datasource guide] for more details.

Reviewer 2: This probably deserves to be explained before you talk about `quarkus test`, because people will wonder: wait, how could tests work, I don't have a database yet -- damn, I need to install one, I'll do that first...

== Simple SQL query

The simplest thing you can do to access your database is to write native SQL queries directly. Create a file called `BookQueries.java` in the package `org.acme`:

[source,java]
----
package org.acme;

import org.hibernate.annotations.processing.SQL;

import java.math.BigDecimal;
import java.util.List;

public interface BookQueries { // <1>

    record BookRecord(Long id, String title, String isbn, BigDecimal price) {} // <2>

>     @SQL("SELECT id, title, isbn, price FROM Book WHERE title = :title") // <3> <4>

Reviewer: I know this is SQL and we're not trying to improve on it, but if people are actually using this a lot, we could… in Jakarta Data, and simplify this to `FROM Book WHERE title = :title` because that first select part can totally be inferred from the return type, just like it does for `@Query`.

WDYT @gavinking ? Should I open a JD issue for this for the future?

    List<BookRecord> findByTitle(String title); // <5>

>     @SQL("SELECT id, title, isbn, price FROM Book WHERE proce < :maxPrice")

Reviewer: Might want to add a footnote here because that I did not think that typo was intentional on my first read.

Author: Moved this part the SQL tutorial instead

    List<BookRecord> cheaperThan(BigDecimal maxPrice);
}
----
<1> A repository interface that groups all your query methods for books. The annotation processor generates the implementation for you.
<2> A Java record that maps to the columns returned by the SQL query. Each field name must match a column name in the result set.
<3> `@SQL` sends raw SQL to the database. The `:title` syntax is a named parameter, it avoids SQL injection attacks.

> <4> `@SQL` is a Hibernate-specific annotation. Jakarta Persistence 4.0 will introduce a standard https://github.com/jakartaee/persistence/pull/992[`@NativeQuery`] equivalent.

Reviewer 2: No need to explain this in a tutorial IMO. This comment is really for us.

```suggestion
// TODO Replace @SQL with Jakarta Persistence 4.0's standard https://github.com/jakartaee/persistence/pull/992[`@NativeQuery`] equivalent.
```

Reviewer: Well it is a warning to users that this will change.

> <5> The method signature defines the parameters. The annotation processor generates the implementation.

Reviewer: ```suggestion
<5> The method signature defines the parameters. 
```
The last part was already said at `<1>`

    assertEquals(4, books.size()); // Clean Code (40), Head First Design Patterns (38), The Pragmatic Programmer (42), Kotlin in Action (44)
}
----

The console in `quarkus test` will show

[source,shell]
----
org.hibernate.exception.SQLGrammarException: JDBC exception executing SQL [ERROR: column "proce" does not exist

Hint: Perhaps you meant to reference the column "book.price".

Position: 26] [

SELECT id, title, isbn, price FROM Book WHERE proce < ?
----

This test will fail at runtime with a SQL error because the column `proce` doesn't exist.
Did you initially spot the typo in `cheaperThan`? The column is called `price`, not `proce`, but the compiler won't tell you.
You're probably going to have integration tests to catch that, but it would be faster if the compiler could validate it earlier.

You can fix the test by modifying the `BookQueries.java` file:

[source,java]
----
    @SQL("SELECT id, title, isbn, price FROM Book WHERE price < :maxPrice") // <1>
    List<BookRecord> cheaperThan(BigDecimal maxPrice);
----
<1> Change `proce` to `price` here

and then all the tests will now pass.

> Writing SQL by hand makes sense for very simple scenarios, or when you need database-specific features such as window functions, recursive functions and similar advanced patterns.

Reviewer 2: I think that's a fine point to make in a talk, but:

1. It's odd to give that kind of lecture in a getting started tutorial. People are here to get started, fast, and they already want to use Quarkus Data. They do not need to be convinced they should use it. IMO this would find a better home in whatever "main" page we decide to use for the Quarkus Data documentation, because *there* we need to explain to people what to use, and why. That page cannot be a tutorial IMO, it needs to be short and to the point.
2. The last statement is antagonizing a whole class of advanced users who prefer to write SQL by hand, just because they know SQL and want control over it -- essentially, people who like Spring Data JDBC. I don't think it's a good idea to tell them they are wrong, because at the core this is a matter of taste, and you will never convince them.

Reviewer: Well, there's people who choose this for taste, we can't save those people. 

But most people will pick SQL only for really good reasons, and we should open this guide with those reasons: why should they use SQL? And equally importantly: when should they not use SQL? And point them to the other tutorial in case they got lost here for no good reason.

And a similar link back from the other tutoral in case they _need_ (or want) SQL.

Author: Removed this part from this tutorial, moved it to the SQL tutorial

== Entities and repositories

We can have a faster development model with the help of the compiler and avoid all these problems by mapping our table to a Java class and defining a repository for it.
We now introduce the concept of an Entity: a Java class that maps to a database table.

Create a file called `Book.java` in the package `org.acme`:

[source,java]
----
package org.acme;

import io.quarkus.hibernate.panache.PanacheRepository;
import io.quarkus.hibernate.panache.WithId;
import jakarta.data.repository.Find;
import jakarta.data.repository.Query;
import jakarta.persistence.Entity;

import java.util.List;

@Entity // <1>

> public class Book extends WithId.AutoLong { // <2>

Reviewer: ```suggestion
public class Book extends ManagedEntity { // <2>
```
This is mixing the future packages with the past type names

    public String title; // <3>
    public String isbn;
    public BigDecimal price;

>     public interface Repo extends PanacheRepository<Book> { // <4>

Reviewer: ```suggestion
    public interface Repo extends ManagedRepository<Book> { // <4>
```

        @Find
        List<Book> findByTitle(String title); // <5>

        @Query("where price < :maxPrice order by title")
        List<Book> cheaperThan(BigDecimal maxPrice); // <6>
    }
}
----
<1> Tells Hibernate that this class maps to a database table. Hibernate will use the class name as the table name and the field names as column names.
<2> Gives the entity a generated `Long` id field automatically.
<3> These public fields map to columns in the `Book` table. Hibernate reads and writes these fields directly.
<4> Defines a repository: a place to put all your query methods for `Book`. By nesting it inside the entity, the queries and the data they operate on live together. You can name this interface whatever you want: `Queries`, `Repo`, `BookQueries`, etc.
<5> `@Find` tells Quarkus Data Hibernate to generate a query based on the method's parameter names. `findByTitle(String title)` becomes a query that matches the `title` field on `Book`. Parameters are inferred from the method signature.

> <6> `@Query` lets you write an HQL fragment when you need more control. It avoids the need to write `select b from Book`. The entity type is inferred from the return type `List<Book>`.

Reviewer: ```suggestion
<6> `@Query` lets you write an HQL fragment when you need more control. It avoids the need to write what you select. The entity type is inferred from the return type `List<Book>`.
```

Otherwise, it would be `from Book` or `select b from Book b`

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@QuarkusTest
public class BookRepositoryTest {

    @Inject
    Book.Repo bookRepo;

    @Test
    void shouldFindBookByTitle() {
        List<Book> books = bookRepo.findByTitle("Effective Java");
        assertEquals(1, books.size());
        assertEquals("Effective Java", books.getFirst().title);
    }

    @Test
    void shouldFindCheaperBooks() {
        List<Book> books = bookRepo.cheaperThan(BigDecimal.valueOf(45));
        assertEquals(4, books.size()); // Clean Code (40), Head First Design Patterns (38), The Pragmatic Programmer (42), Kotlin in Action (44)
    }

    @Test
    void shouldReturnEmptyListForUnknownTitle() {
        List<Book> books = bookRepo.findByTitle("Unknown Book");
        assertTrue(books.isEmpty());
    }
}
----

Now if you make the same kind of typo we had earlier, the build fails. Try changing the parameter name to:

[source,java]
----
        @Find
        List<Book> findByTitle(String tilte);
----

You should see a compilation error like:

[source]
----
error: no matching field named 'tilte' in entity class 'org.acme.Book'
----

> Get the parameter back to the correct name so that we can proceed further.

Reviewer: This is the hardest I've worked following a tutorial 😓

== Creating new entities

> So far we've been reading data, but our bookstore is crowdsourced: users can add books they want to share.

Reviewer 2: Not sure you need to build a story for this. We just need to persist data, period -- you don't have to justify yourself :)

Author: Removed, though I personally find domain examples make tutorials more engaging than abstract ones

We need to persist new books to the database.

> Let's update `Book` to implement `PanacheEntity.Stateless`, which gives us lifecycle operations directly on the entity:

Reviewer 2: I would do this from the start, because this step unnecessarily complicates your tutorial. We're already straying quite far from the "straight to the point" approach in [the Spring Data JPA getting started guide](https://docs.spring.io/spring-data/jpa/reference/jpa/getting-started.html).

You can just omit explaining the lifecycle operations until you get to this section, and explain them right when you show how to implement the write operation.

[source,java]
----
@Entity

> public class Book extends WithId.AutoLong implements PanacheEntity.Stateless { // <1>

Reviewer: ```suggestion
public class Book extends RecordEntity { // <1>
```

    public String title;
    public String isbn;
    public BigDecimal price;

>     public interface Repo extends PanacheRepository.Stateless<Book, Long> { // <2>

Reviewer: ```suggestion
    public interface Repo extends RecordRepository<Book> { // <2>
```

        @Find
        List<Book> findByTitle(String title);

        @Query("where price < :maxPrice order by title")
        List<Book> cheaperThan(BigDecimal maxPrice);
    }
}
----
<1> This is the new part: `implements PanacheEntity.Stateless` gives our `Book` class methods to create, delete or update books without the need to use specific methods in repositories.

<2> The repository also changed to `PanacheRepository.Stateless<Book, Long>`.

> // TODO: PanacheRepository<Book> infers Long, but Stateless/Managed/Reactive all require <Book, Long> explicitly. We should add single-parameter overloads (defaulting to Long) for consistency.

Reviewer: This is fixed in the rename branch, they all default to `Long` unless `.CustomId`

        List<Book> books = bookRepo.findByTitle("New Book");
        assertEquals(1, books.size());
        assertEquals("978-1111111111", books.getFirst().isbn);
        assertEquals(BigDecimal.valueOf(29.99), books.getFirst().price);
    }
}
----
<1> `@TestTransaction` rolls back all database changes when the test method ends. Without it, the inserted book would stay in the database and accumulate on every re-run, causing assertions on counts or prices to fail. Use it on any test that writes data.

[NOTE]
====
If you're an experienced Hibernate user, you'll find this part very different from what you're used to.
When you call `insert()`, `update()`, or `delete()`, an operation is instantly triggered against the database, without calling `.flush()` explicitly.
Managed entities are nowhere to be seen, and neither is the idea of detached and persisted objects.
This is intentional: it's a simpler model that uses `StatelessSession` under the covers.
In this mode every change to an object requires an explicit save operation.
If you're not interested in this style and want the full power of managed entities, skip ahead to <<managed-session>>.

> ====

Reviewer 2: This complicates the tutorial. Leave it for the section about managed sessions?

Reviewer: I'm not sure. It explains to Hibernate users why this is different and tells them where to find what they're used to elsewhere.  Perhaps it's too long in details. Perhaps it should be titled "note for hibernate frequent users" (not sure about what term to use here).

=== About entity identifiers

> `WithId.AutoLong` is more powerful than Hibernate's traditional `@Id @GeneratedValue` pair: there are variants for different ID types out of the box:

Reviewer 2: It is not more powerful, quite obviously, since it's based on Hibernate ORM :)

It's convenient syntactical sugar.

- `WithId.AutoLong`, a generated `Long` id
- `WithId.AutoUUID`, a generated `UUID` id
- `WithId.AutoString`, a generated `String` id

Of course you can always use `@Id` and `@GeneratedValue` directly if you prefer, or if you need a custom ID strategy.

> See the xref:quarkus-data-hibernate.adoc#_about_entity_identifiers[reference guide] for the full list of options.

Reviewer 2: IMO this should just be a link to the reference documentation for Quarkus Data in your code example about writing an entity. E.g. on `extends WithId.AutoLong`, you say `this defines IDs as auto-generated longs; other types of IDs have convenient shortcuts or advanced customization options, see xref:quarkus-data-reference.adoc#ids[this section of the reference]`.

Of course, that requires that we have a Quarkus Data reference in the first place :)

Reviewer: There's quite a section about custom IDs already in the Quarkus Data guide (reference or not), so you could resume this entire section into the pointer.

[[managed-session]]
== Managed session

In our bookstore we want to run a promotional sale: apply a discount to all books in a given country, but only if the publisher allows it.
Some publishers offer discounts, others don't.

Our database already has publishers (we set them up in `import.sql`), but we haven't mapped them as entities yet. Let's do that now. Create `Publisher.java`:

[source,java]
----
package org.acme;

import io.quarkus.hibernate.panache.PanacheRepository;
import io.quarkus.hibernate.panache.WithId;
import jakarta.persistence.Entity;

@Entity

> public class Publisher extends WithId.AutoLong {

Reviewer: ```suggestion
public class Publisher extends RecordEntity {
```

    public String name;
    public String country;
    public boolean allowsDiscounts;

>     public interface Repo extends PanacheRepository.Stateless<Publisher, Long> {

Reviewer: ```suggestion
    public interface Repo extends RecordRepository<Publisher> {
```

    }
}
----

And add a `publisher` relationship to `Book`, plus a stateless repository and a query to find books by publisher country:

[source,java]
----
@Entity

> public class Book extends WithId.AutoLong implements PanacheEntity.Stateless {

Reviewer: ```suggestion
public class Book extends StatelessEntity {
```

    public String title;
    public String isbn;
    public BigDecimal price;
    @ManyToOne(fetch = FetchType.LAZY) // <1>
    public Publisher publisher;

>     public interface Repo extends PanacheRepository.Stateless<Book, Long> {

Reviewer: ```suggestion
    public interface Repo extends StatelessRepository<Book> {
```

        @Find
        List<Book> findByTitle(String title);

        @Query("where price < :maxPrice order by title")
        List<Book> cheaperThan(BigDecimal maxPrice);

        @Query("where publisher.country = :country")
        List<Book> findByPublisherCountry(String country);
    }
}
----

> <1> We use `FetchType.LAZY` so that loading a book doesn't automatically join and fetch its publisher.

Reviewer 2: This is kind of odd. Lazy to-ones are not that frequent in my experience. Neither is explicit configuration of laziness, btw -- defaults are generally fine, and probably should be in such a getting started tutorial.

You could find some other example with a to-many associations? That way you don't even need to mention laziness here, and can mention it further down.

Reviewer: Indeed, I thought fetches were explicit in stateless sessions?

Author: Changed from ManyToOne to OneToMany so it's lazy by default

Now let's implement our discount service:

[source,java]
----
@ApplicationScoped
public class SalesService {

    @Inject
    Book.Repo bookRepo;

    @Inject
    Publisher.Repo publisherRepo;

    @Transactional
    public void applyCountrySale(String country, BigDecimal discountPercent) {
        List<Book> books = bookRepo.findByPublisherCountry(country);
        List<Book> modified = new ArrayList<>(); // <1>
        BigDecimal factor = BigDecimal.ONE.subtract(discountPercent.divide(BigDecimal.valueOf(100)));
        for (Book book : books) {
            Publisher publisher = publisherRepo.findById(book.publisher.id); // <2>
            if (publisher.allowsDiscounts) {
                book.price = book.price.multiply(factor).setScale(2, RoundingMode.HALF_UP);
                modified.add(book); // <3>
            }
        }

>         // Other business logic that modifies books i.e. applying loyalty points, rounding prices ...

Reviewer: This is a contrived example, because in practice we're only updating each book once.

Author: Yes that's the whole point. Update each book only once, and only those that were actually changed

        // Since we want to update them only once, we collect them to avoid multiple updates within the loops
        for (Book book : modified) {
            bookRepo.update(book); // <4>
        }
    }
}
----
<1> We collect the books we modify so we can update them all at the end, after all business logic has run. Updating inside the body will result in multiple SQL update statements.

> <2> In a stateless session, lazy associations are not fetched automatically. We must load the publisher explicitly. A join fetch query in the Book repository could be used instead of a separate fetch, but for the scope of the tutorial we're not introducing join fetches.

Reviewer 2: IMO you're getting too far here. You're addressing beginners -- if you think join fetches are too complicated to explain here, you shouldn't mention them. Also, beginners might not even know what a lazy association is, so we'd need to explain that.

Perhaps more importantly, this is rather bad code. In fact, it's textbook N+1 select.

I think it would make more sense to explain how one would *properly* implement this with a stateless session. If you keep a ToOne association here, it would need to be eager, and if you replace it with a ToMany association, it would make sense to explain join fetches, because they would be a core feature for any application using stateless session.

Author: Replaced with ToMany association

<3> We add the book to our list of modified objects.
<4> We update once at the end, after all modifications are done.

> Notice how the persistence concerns, tracking which objects changed, fetching lazy associations, explicitly calling `update()`, are mixed in with the business logic. This method is beginning to become complex.

Reviewer: ```suggestion
Notice how the persistence concerns, tracking which objects changed, fetching lazy associations, explicitly calling `update()`, are mixed in with the business logic. This method is becoming complex.
```

import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

@QuarkusTest
public class SalesServiceTest {

    @Inject
    SalesService salesService;

    @Inject
    Book.Repo bookRepo;

    @Test
    @TestTransaction // <1>
    void shouldApplySaleOnlyToPublishersThatAllowDiscounts() {
        salesService.applyCountrySale("US", BigDecimal.valueOf(10));

        // Addison-Wesley (US, allows discounts): prices reduced by 10%
        assertEquals(new BigDecimal("40.50"), bookRepo.findByTitle("Effective Java").getFirst().price); // was: 45.00
        assertEquals(new BigDecimal("36.00"), bookRepo.findByTitle("Clean Code").getFirst().price); // was: 40.00

        // Prentice Hall (US, no discounts): prices unchanged
        assertEquals(new BigDecimal("55.00"), bookRepo.findByTitle("Domain-Driven Design").getFirst().price);
        assertEquals(new BigDecimal("50.00"), bookRepo.findByTitle("Refactoring").getFirst().price);
    }
}
----
<1> Rolls back the discounted prices after the test, so other tests and re-runs always see the original data from `import.sql`.

This works, but it can be simpler.

> Quarkus Data Hibernate supports a "managed persistence context" that tracks objects for you: Hibernate monitors every entity it loads, detects which fields changed, and generates the minimal SQL needed to keep entities synchronized with the database at commit time. You don't need to track dirty objects, and lazy associations are fetched automatically when you access them.

Reviewer: I feel like this is a very long introduction where we show a lot of code just to justify the existence of managed sessions.

I think we should rather make a list of pros/cons of each session type. That would be more informative and to the point, and remove a lot of code that is intended to be thrown away.

Switch `Book` to use `PanacheEntity.Managed` and a managed repository:

[source,java]
----
@Entity

> public class Book extends WithId.AutoLong implements PanacheEntity.Managed { // <1>

Reviewer: ```suggestion
public class Book extends ManagedEntity { // <1>
```

    public String title;
    public String isbn;
    public BigDecimal price;
    @ManyToOne(fetch = FetchType.LAZY)
    public Publisher publisher;

>     public interface Repo extends PanacheRepository<Book> { // <2>

Reviewer: ```suggestion
    public interface Repo extends ManagedRepository<Book> { // <2>
```

        Book book = new Book();
        book.title = title;
        book.isbn = isbn;
        book.price = price;
        book.persist(); // <1>
    }
}
----
<1> `persist()` replaces `insert()`. Unlike `insert()`, it doesn't execute SQL immediately, Hibernate will flush the change when the transaction commits.

Now the service becomes pure business logic:

[source,java]
----
@ApplicationScoped
public class SalesService {

    @Inject
    Book.Repo bookRepo;

    @Transactional
    public void applyCountrySale(String country, BigDecimal discountPercent) {
        BigDecimal factor = BigDecimal.ONE.subtract(discountPercent.divide(BigDecimal.valueOf(100)));
        List<Book> books = bookRepo.findByPublisherCountry(country);
        for (Book book : books) {
            if (book.publisher.allowsDiscounts) { // <1>
                book.price = book.price.multiply(factor).setScale(2, RoundingMode.HALF_UP); // <2>
            }
        }
    } // <3>
}
----
<1> Accessing `book.publisher` triggers a lazy fetch: Hibernate loads the publisher from the database automatically with a query.
<2> We just set the field. No need to call `update()`.

> <3> When the `@Transactional` method returns, Hibernate compares every managed entity against its original state and generates the correct SQL. Books whose publisher didn't allow discounts were never changed, so Hibernate skips them entirely.

Reviewer: I think this is not true, and scary: it does not compare every managed entity, I think it tracks those that were modified, no? If true, this is less scary because obviously more efficient.

Reviewer 2: It's not true in Quarkus because we use bytecode enhancement, which allows inline dirty-tracking.

It's true in vanilla Hibernate ORM: that's called dirty-checking.

Reviewer: You know, I never realised these two terms were different :-/

The test is exactly the same: the behavior hasn't changed, just the implementation got simpler.

> It's up to you to decide whether to use a Managed Entity or a Stateless one. The Managed Repository is more expressive, but reasoning about it is more complex and might require knowing some details about how Hibernate handles objects. If you're an experienced Hibernate developer, you probably already know the discussions about n+1 queries, and you expect that changing an object's field doesn't immediately hit the database. Some users prefer a more explicit way to control the database. Either way is fine, and Quarkus Data Hibernate makes it easy to use both and to switch without committing to one style or the other.

Reviewer 2: ```suggestion
The test is exactly the same: the behavior hasn't changed, the implementation only got simpler.

It's up to you to decide whether to use a Managed Entity or a Stateless one. The Managed Repository is more expressive, but reasoning about it is more complex and might require knowing some details about how Hibernate handles objects. If you're an experienced Hibernate developer, you probably already know the discussions about n+1 queries, and you expect that changing an object's field doesn't immediately hit the database. Some users prefer a more explicit way to control the database. Either way is fine, and Quarkus Data Hibernate makes it easy to use both and to switch without committing to one style or the other.
```

Reviewer: I'm not sure that N+1 issues are related to "not immediately hitting the DB" and this may appear to suggest that 🤔

First, add the reactive dependencies:

[source,bash]
----
quarkus extension add quarkus-hibernate-reactive
quarkus extension add quarkus-hibernate-reactive-panache
quarkus extension add quarkus-reactive-pg-client
----

NOTE: `quarkus-hibernate-reactive-panache` is needed because `quarkus-hibernate-reactive` does not yet transitively include the classes required by the annotation processor to generate reactive repository implementations (see https://github.com/quarkusio/quarkus/issues/54892[#54892]).

`quarkus-hibernate-reactive` adds the reactive Hibernate engine, which uses non-blocking I/O under the covers instead of JDBC.
`quarkus-reactive-pg-client` is the reactive equivalent of `quarkus-jdbc-postgresql`. It provides a non-blocking PostgreSQL driver built on Vert.x. As its blocking counterpart, it's mandatory for Quarkus Data Hibernate to work.

Now let's change only the minimal part of the code you need to try reactive. Let's add a second nested repository for reactive operations:

[source,java]
----
@Entity

> public class Book extends WithId.AutoLong implements PanacheEntity.Managed {

Reviewer: ```suggestion
public class Book extends ManagedEntity {
```

    public String title;
    public String isbn;
    public BigDecimal price;
    @ManyToOne(fetch = FetchType.LAZY)
    public Publisher publisher;

>     public interface Repo extends PanacheRepository<Book> { // <1>

Reviewer: ```suggestion
    public interface Repo extends ManagedRepository<Book> { // <1>
```

        @Find
        List<Book> findByTitle(String title);

        @Query("where price < :maxPrice order by title")
        List<Book> cheaperThan(BigDecimal maxPrice);

        @Query("where publisher.country = :country")
        List<Book> findByPublisherCountry(String country);
    }

>     public interface ReactiveRepo extends PanacheRepository.Reactive<Book, Long> { // <2>

Reviewer: ```suggestion
    public interface ReactiveRepo extends ManagedRepository.Reactive<Book> { // <2>
```

import io.quarkus.test.vertx.RunOnVertxContext;
import io.quarkus.test.vertx.UniAsserter;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

@QuarkusTest
public class ReactiveBookRepositoryTest {

    @Inject
    Book.ReactiveRepo reactiveBookRepo;

    @Test
    @RunOnVertxContext // <1>
    @TestReactiveTransaction // <2>
    void shouldFindBookByTitleReactively(UniAsserter asserter) {
        asserter.assertThat( // <3>
            () -> reactiveBookRepo.findByTitle("Effective Java"),
            books -> {
                assertEquals(1, books.size());
                assertEquals("Effective Java", books.getFirst().title);
            }
        );
    }

    @Test
    @RunOnVertxContext
    @TestReactiveTransaction
    void shouldReturnEmptyListForUnknownTitleReactively(UniAsserter asserter) {
        asserter.assertThat(
            () -> reactiveBookRepo.findByTitle("Unknown Book"),
            books -> assertTrue(books.isEmpty())
        );
    }
}
----
<1> `@RunOnVertxContext` ensures the test runs on the Vert.x event loop, which is required for reactive database access.

> <2> `@TestReactiveTransaction` opens a reactive session and transaction for the test, and rolls back at the end.

Reviewer 2: Not about this tutorial per se, but... this is awfully complicated. In an ideal world we'd require only `@TestTransaction` and `@RunOnVertxContext` would be inferred from `UniAsserter`. No?

cc @FroMage

Reviewer: Absolutely, we should improve this.

Reviewer: Come to think about it, I wonder if we should not use the same sort of mechanism as I recently did to allow returning a `Uni` from `@Startup` and just plug it in the vertx loop. That would make for much nicer test methods.
